File: 876-middle-of-the-linked-list/876-middle-of-the-linked-list.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn middle_node(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let (mut fast, mut slow, mut i) = (head.clone(), head.clone(), 0);
        
        while fast.is_some() {
            if(i % 2 == 1) {
                slow = slow.unwrap().next;
            }
            fast = fast.unwrap().next;
            i+=1;
        }
        
        slow
    }
}

File: 876-middle-of-the-linked-list/876-middle-of-the-linked-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function middleNode(head: ListNode | null): ListNode | null {
    if (!head) return null;
    let fast = head;
    let slow = head;
    while (fast) {
        fast = fast.next;
        if (fast) {
            fast = fast.next;
            slow = slow.next;
        }
    }
    return slow;
};